Feature/faster normal space filter

jsburklund opened the pull request:
Improves the runtime performance of the normal space filter. Replaces rejection sampling with an equivalent, but more efficient method. Improves runtime determinism as well.  Speed up is from about 3-4 seconds on 35k points down to tens of msecs.

Commit:
Remove trailing whitespace

Commit:
Improve speed of Normal Space filtering.

Use more efficient sampling method instead of
rejection sampling. Improves runtime speed and
determinism.

remod commented:
I don't know exactly how this filter works, but I am trusting you :)

jsburklund commented:
The previous method used rejection sampling with binned normal values. When the data is filtered, the normals are put into specific bins based on their angles. The previous method would first select a random bin, and test to see if the bin was empty. If the bin was empty, another random value was sampled from the random generator until it found a bin that had samples in it. After selecting the bin to use, it sampled a random point from the bin that hadn't already been sampled. If the selected point had already been sampled, another random point was selected until one was found that hadn't already been sampled. In an environment with very sparse bins, like a rectangular room, the previous method had to sample tens of billions of random values from the random generator in order to filter 35000 points. The problem was with trying to sample low-probability events with sparse bins, and bins that had most of the elements already selected.

The new method removes the rejection based sampling, and instead only samples from the available bins. When data first comes in, the data is put into the appropriate bin in a random order by generating a randomly shuffled sequence of indices.  After putting data into the bins, the empty bins are removed. When filtering, only the list of available bins is randomly selected from. When a bin is selected, the last element in the list of points in the bin is popped which should be "random" since elements were originally loaded in a random order into the bin. When the last element from a bin is removed, the bin is removed from the list of available bins.

This improves determinism by using a constant number of random accesses into binned data.  There are likely some other performance and determinism improvements to be made by preallocating pools for the vectors, but the filter is good enough to run reasonably now.

YoshuaNava commented:
Hey,
Thanks for the contribution. This looks like a really nice improvement.

So overall you:
1) Modernized the loop that builds the sampling space. You first shuffle randomly the indices of the point cloud points, and then run a range-based for loop over them.
2) Moved away from sampling in the normal space, to sampling inside a discretized space, where you know a sample will always hit a bucket.
3) Popped indices that you already sampled and chose to keep in the filtered cloud, to prevent going over them again.

It sounds quite solid. Have you been able to check how the results look?

Additionally, I think it would make sense to contribute this to the upstream Libpointmatcher repo.

jsburklund commented:
The sampling was already in the discretized normal space. The only difference is that any bucket that is randomly selected is guaranteed to have points that could still be sampled.

When selecting a random point inside of a particular bucket, the points are already inserted into the bucket in a random order. Rather than randomly selecting a point from the list of points still available in the bucket and removing it (which would have O(N) cost for removing from the middle of a vector), the last element is simply popped which is a O(1) cost operation.  This isn't exactly the same as randomly selecting an available point each time, but should be close enough if the points are inserted in a random order to begin with.

samuelba commented:
At some point we should sync again with the original repository from ASL. And try to get this change into the original repo. I think they did quite some improvements regarding pointers etc.

Commit:
Remove trailing whitespace

Commit:
Clamp normal components to [-1, 1]

Approximation errors sometimes cause the components of the normal
vector to be beyond the range of [-1, 1]. Normal components are
now clamped to prevent computation errors downstream.

Commit:
Add assertion checks for debugging

Commit:
Fix hashing issues.

Explicity use the correct rounding direction so that enough
buckets are allocated, and handle wrapping theta appropriately
at Pi. Previously, if theta=Pi, the hash returned an index out
of bounds of the preallocated memory.

remod commented:
@YoshuaNava Are you willing to accept? :)

YoshuaNava commented:
Hi @remod,
I thought after Samuel's comment we would be submitting these changes upstream. Nonetheless, if we are gonna merge them here first it also works.

remod commented:
I created a Wrike task for synchronizing the two repos.
